tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp

```diff
@@ -0,0 +1,143 @@
+    constexpr uint32_t in0_mcast_num_cores                = get_compile_time_arg_val(5);
+    // block args
+    constexpr uint32_t num_blocks                         = get_compile_time_arg_val(6);
+    // in0 mcast args
+    constexpr uint32_t in0_mcast_dest_noc_start_x         = get_compile_time_arg_val(7);
+    constexpr uint32_t in0_mcast_dest_noc_start_y         = get_compile_time_arg_val(8);
+    constexpr uint32_t in0_mcast_dest_noc_end_x           = get_compile_time_arg_val(9);
+    constexpr uint32_t in0_mcast_dest_noc_end_y           = get_compile_time_arg_val(10);
+    // in0 semaphore always valid
+    constexpr uint32_t in0_mcast_sender_valid_semaphore   = get_compile_time_arg_val(11);
+
+    constexpr uint32_t num_blocks_per_shard               = get_compile_time_arg_val(12);
+    constexpr uint32_t num_storage_cores = num_blocks / num_blocks_per_shard;
+
+    // RUNTIME ARGS
+    const bool is_worker_core                                     = get_arg_val<uint32_t>(0) == 1;
+    const uint32_t sender_id                                      = get_arg_val<uint32_t>(1);
+    volatile tt_l1_ptr uint32_t * in0_mcast_sender_noc_x          = (volatile tt_l1_ptr uint32_t*)(get_arg_addr(2));
+    volatile tt_l1_ptr uint32_t * in0_mcast_sender_noc_y          = (volatile tt_l1_ptr uint32_t*)(get_arg_addr(2 + num_storage_cores));
+
+    const uint32_t sender_block_id = sender_id * num_blocks_per_shard;
+
+    constexpr uint32_t cb_id_in0 = 0;
+    constexpr uint32_t cb_id_in2 = 2; // Sharded cb
+
+
+    const uint32_t in0_single_tile_size_bytes = get_tile_size(cb_id_in0);
+    const DataFormat in0_data_format = get_dataformat(cb_id_in0);
+
+
+    uint32_t l1_write_addr_in0;
+
+    // Set ur local VALID value, to be mcasted to destinations flag address after the data has been mcasted
+    volatile tt_l1_ptr uint32_t* in0_mcast_receiver_semaphore_addr_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(in0_mcast_receiver_semaphore_addr);
+    *(in0_mcast_receiver_semaphore_addr_ptr) = VALID;
+    // local address that will be atomically incremented by mcast receivers, to know when all receivers are ready
+    // to receive the mcast
+    volatile tt_l1_ptr uint32_t* in0_mcast_sender_semaphore_addr_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(in0_mcast_sender_semaphore_addr);
+
+    const uint64_t in0_mcast_receiver_semaphore_noc_addr = get_noc_multicast_addr(
+        in0_mcast_dest_noc_start_x,
+        in0_mcast_dest_noc_start_y,
+        in0_mcast_dest_noc_end_x,
+        in0_mcast_dest_noc_end_y,
+        in0_mcast_receiver_semaphore_addr);
+
+    const uint64_t in0_multicast_data_noc = get_noc_multicast_addr(
+        in0_mcast_dest_noc_start_x,
+        in0_mcast_dest_noc_start_y,
+        in0_mcast_dest_noc_end_x,
+        in0_mcast_dest_noc_end_y,
+        0);
+
+    uint32_t local_read_addr = get_read_ptr(cb_id_in2);
+
+    if (not is_worker_core) {
+
```

Comment:
We should `constexpr` all paths like this

Reply:
it is run-time arg, can't use